data-plane/src/acme/order.rs

```diff
@@ -0,0 +1,304 @@
+        let body_str = from_utf8(&resp_bytes)?;
+        let mut order: Order<_> = serde_json::from_str(body_str)?;
+
+        order.account = Some(account.clone());
+        order.url = self.url.clone();
+        Ok(order)
+    }
+
+    pub async fn certificate(&self) -> Result<Option<Vec<X509>>, AcmeError> {
+        let certificate_url = match self.certificate_url.clone() {
+            Some(certificate_url) => certificate_url,
+            None => return Ok(None),
+        };
+
+        let (account, directory) = self.get_account_and_directory()?;
+
+        let response = directory
+            .authenticated_request(
+                &certificate_url,
+                "POST",
+                None,
+                &account
+                    .private_key
+                    .clone()
+                    .ok_or(AcmeError::FieldNotFound("private_key".into()))?,
+                &Some(account.id.clone()),
+            )
+            .await?;
+
+        let body_bytes = hyper::body::to_bytes(response.into_body()).await?;
+
+        Ok(Some(X509::stack_from_pem(&body_bytes)?))
+    }
+
+    pub async fn poll(&self) -> Result<Order<T>, AcmeError> {
+        let (account, directory) = self.get_account_and_directory()?;
+
+        let response = directory
+            .authenticated_request(
+                &self.url,
+                "POST",
+                None,
+                &account
+                    .private_key
+                    .clone()
+                    .ok_or(AcmeError::FieldNotFound("private_key".into()))?,
+                &Some(account.id.clone()),
+            )
+            .await?;
+
+        let resp_bytes = hyper::body::to_bytes(response.into_body()).await?;
+        let body_str = from_utf8(&resp_bytes)?;
+
+        let mut order: Order<_> = serde_json::from_str(body_str)?;
+
+        order.account = Some(account.clone());
+        order.url = self.url.clone();
+        Ok(order)
+    }
+
+    pub async fn wait_ready(
+        self,
+        poll_interval: Duration,
+        attempts: usize,
+    ) -> Result<Order<T>, AcmeError> {
+        let mut order = self;
+
+        let mut i: usize = 0;
+
+        while order.status == OrderStatus::Pending {
+            if i >= attempts {
+                return Err(AcmeError::General(
+                    "Max attempts reached for polling order.".into(),
+                ));
+            }
+            tokio::time::sleep(poll_interval).await;
+            order = order.poll().await?;
+            i += 1;
+        }
+
+        Ok(order)
+    }
+
+    pub async fn wait_done(
+        self,
+        poll_interval: Duration,
+        attempts: usize,
+    ) -> Result<Order<T>, AcmeError> {
+        let mut order = self;
+
+        let mut i: usize = 0;
+
+        while order.status == OrderStatus::Pending
+            || order.status == OrderStatus::Ready
+            || order.status == OrderStatus::Processing
```

Comment:
Worth moving this into a function on the OrderStatus enum?

Reply:
Yeah good shout!